README
youtube-dl-playlist is a Node.js program that uses youtube-dl to download an entire playlist loaded in from a JSON file.
The program automatically will download all tracks with the best audio available, then convert them into mp3 files and
add ID3v2 frames to the file.

The JSON file must be an array where each item must match the following schema:
{
	"title": string,
	"artist": string,
	"featured_artists"?: string[],
	"album"?: string,
	"nr"?: number,
	"year"?: number,
	"comments"?: number | string[],
	"url": string,
	"fallback_urls"?: string[],
	"other"?: any
}


CHANGELOG.md
<!--
  Copyright (c) 2023 Michael Federczuk
  SPDX-License-Identifier: CC-BY-SA-4.0
-->

<!-- markdownlint-disable no-duplicate-heading -->

# Changelog #

All notable changes to this project will be documented in this file.
The format is based on [**Keep a Changelog v1.0.0**](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [**Semantic Versioning v2.0.0**](https://semver.org/spec/v2.0.0.html).

## Unreleased ##

### Added ###

* New option `--only-show-errors`

## [v1.0.0-indev08] - 2021-05-22 ##

[v1.0.0-indev08]: https://github.com/mfederczuk/youtube-dl-playlist/releases/tag/v1.0.0-indev08

### Changed ###

* Checks for **cURL** and **FFmpeg** in the `PATH` are now only done when they are actually needed to download tracks

### Fixed ###

* Tracks downloaded with **cURL** would "fail" when the file was already an mp3 file

## [v1.0.0-indev07] - 2021-05-13 ##

[v1.0.0-indev07]: https://github.com/mfederczuk/youtube-dl-playlist/releases/tag/v1.0.0-indev07

### Fixed ###

* Tracks with titles that do not contain any letters or digits would fail to download

## [v1.0.0-indev06] - 2021-02-24 ##

[v1.0.0-indev06]: https://github.com/mfederczuk/youtube-dl-playlist/releases/tag/v1.0.0-indev06

### Added ###

* `"file"`, `"fallback_files"` track fields.  
  These URLs will be downloaded with **cURL** instead of **youtube-dl**

## [v1.0.0-indev05] - 2021-02-15 ##

[v1.0.0-indev05]: https://github.com/mfederczuk/youtube-dl-playlist/releases/tag/v1.0.0-indev05

### Added ###

* `"featured_artists"`, `"comments"` and `"other"` track fields
* The `"comments"` field will now also be written to the ID3v2 tags, along with a base comment.  
  Also, the URL that was used to download the track is now also saved in the ID3v2 tags

## [v1.0.0-indev04] - 2021-02-11 ##

[v1.0.0-indev04]: https://github.com/mfederczuk/youtube-dl-playlist/releases/tag/v1.0.0-indev04

### Changed ###

* Sorting order was slightly changed: album position has higher sorting priority than the title when both tracks have
  albums
* When at least one track fails to download, the program will exit with code `32`

## [v1.0.0-indev03] - 2021-02-11 ##

[v1.0.0-indev03]: https://github.com/mfederczuk/youtube-dl-playlist/releases/tag/v1.0.0-indev03

### Added ###

* Playlist JSON warnings are being displayed

### Fixed ###

* Some processes would never end be when downloading small playlists

## [v1.0.0-indev02] - 2021-02-11 ##

[v1.0.0-indev02]: https://github.com/mfederczuk/youtube-dl-playlist/releases/tag/v1.0.0-indev02

### Fixed ###

* Added a missing babel dependency - was required for the program to run

## [v1.0.0-indev01] - 2021-02-10 ##

[v1.0.0-indev01]: https://github.com/mfederczuk/youtube-dl-playlist/releases/tag/v1.0.0-indev01

Initial release


src/main.ts
/*
 * Copyright (c) 2021 Michael Federczuk
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

import { existsSync, readdirSync, readFileSync, statSync, writeFileSync } from "fs";
import * as path from "path";
import { DeepMutable } from "./mutable";
import { Playlist } from "./playlist";

const scriptName = path.basename(process.argv[1]);
const usage = `usage: ${scriptName} [--download [--high-effort] [--only-how-errors]] [--sort] [--compact|--pretty=(tab|<number_of_spaces>)] <playlist_file>`;

interface CliInput {
	readonly filePath?: string;
	readonly download: (false | {
		readonly highEffort: boolean;
		readonly onlyShowError: boolean;
	});
	readonly rewrite: (false | {
		readonly format?: ("compact" | "tab" | number);
		readonly sort: boolean;
	});
}

const cliInput: CliInput = process.argv.slice(2).reduce((cliInput, arg) => {
	if(cliInput.processOptions && arg.startsWith("--")) {
		if(arg === "--") {
			cliInput.processOptions = false;
		} else if(arg === "--download") {
			if(cliInput.download === false) {
				cliInput.download = { highEffort: false, onlyShowError: false };
			}
		} else if(arg === "--high-effort") {
			if(cliInput.download === false) {
				cliInput.download = { highEffort: true, onlyShowError: false };
			} else {
				cliInput.download.highEffort = true;
			}
		} else if(arg === "--only-show-errors") {
			if(cliInput.download === false) {
				cliInput.download = { highEffort: false, onlyShowError: true };
			} else {
				cliInput.download.onlyShowError = true;
			}
		} else if(arg === "--compact") {
			if(cliInput.rewrite === false) {
				cliInput.rewrite = { format: "compact", sort: false };
			} else {
				cliInput.rewrite.format = "compact";
			}
		} else if(arg === "--sort") {
			if(cliInput.rewrite === false) {
				cliInput.rewrite = { sort: true };
			} else {
				cliInput.rewrite.sort = true;
			}
		} else {
			const match = arg.match(/^--pretty=(tab|\d+)$/);

			if(match === null) {
				console.error(`${scriptName}: ${arg}: invalid option`);
				console.error(usage);
				process.exit(5);
			}

			const optArg = match[1];

			if(optArg === "tab") {
				if(cliInput.rewrite === false) {
					cliInput.rewrite = { format: "tab", sort: false };
				} else {
					cliInput.rewrite.format = "tab";
				}
			} else {
				const spaces = Number(optArg);
				if(cliInput.rewrite === false) {
					cliInput.rewrite = { format: spaces, sort: false };
				} else {
					cliInput.rewrite.format = spaces;
				}
			}
		}
	} else if(cliInput.filePath === undefined) {
		cliInput.filePath = arg;
	} else {
		console.error(`${scriptName}: too many arguments`);
		console.error(usage);
		process.exit(4);
	}

	return cliInput;
}, {
	filePath: undefined,
	download: false,
	rewrite: false,
	processOptions: true
} as (DeepMutable<CliInput> & { processOptions: boolean; }));

if(cliInput.download === false && cliInput.rewrite === false) {
	console.error(`${scriptName}: either the --download, --compact, --pretty or --sort options must be given`);
	console.error(usage);
	process.exit(13);
}

if(cliInput.rewrite !== false && cliInput.rewrite.format === undefined) {
	console.error(`${scriptName}: if the --sort option is given, either the --compact or --pretty options must be given as well`);
	console.error(usage);
	process.exit(14);
}

if(cliInput.filePath === undefined) {
	console.error(`${scriptName}: missing argument: <playlist_file>`);
	console.error(usage);
	process.exit(3);
}

if(!existsSync(cliInput.filePath)) {
	console.error(`${scriptName}: ${cliInput.filePath}: no such file`);
	process.exit(24);
}

const playlistJson = JSON.parse(readFileSync(cliInput.filePath).toString());
const validationResult = Playlist.validateJSON(playlistJson);

if(validationResult.error !== undefined) {
	console.error(`${scriptName}: ${cliInput.filePath}: ${validationResult.error.message}`);
	process.exit(48);
}

if(validationResult.warning !== undefined) {
	console.warn(`${cliInput.filePath}: ${validationResult.warning.message}`);
}

const playlist = validationResult.value;

if(playlist === undefined) {
	console.error(`${scriptName}: ${cliInput.filePath}: error validating playlist JSON`);
	process.exit(48);
}

if(cliInput.rewrite !== false) {
	const obj = (cliInput.rewrite.sort ? playlist.toJSON() : playlistJson);

	let space: (number | "\t" | undefined);
	if(cliInput.rewrite.format === "tab") {
		space = "\t";
	} else if(typeof(cliInput.rewrite.format) === "number") {
		space = cliInput.rewrite.format;
	}

	writeFileSync(cliInput.filePath, JSON.stringify(obj, undefined, space) + (space !== undefined ? "\n" : ""));
}

if(cliInput.download !== false) {
	const needsCurlAndFfmpeg = [...playlist.values()]
		.flatMap((track) => track.urls)
		.some((url) => url.protocol.startsWith("ytdlpl-curl-"));

	let hasYoutubeDl = false,
		hasCurl = false,
		hasFfmpeg = false;

	const pathDirs = process.env["PATH"]?.split(":");

	if(pathDirs instanceof Array) {
		outer: for(const pathDir of pathDirs) {
			if(!existsSync(pathDir) || !statSync(pathDir).isDirectory()) {
				continue;
			}

			for(const entry of readdirSync(pathDir)) {
				if(entry === "youtube-dl") hasYoutubeDl = true;
				if(entry === "curl") hasCurl = true;
				if(entry === "ffmpeg") hasFfmpeg = true;

				if(hasYoutubeDl && hasCurl && hasFfmpeg) {
					break outer;
				}
			}
		}
	}

	[
		[true, hasYoutubeDl, "youtube-dl"],
		[needsCurlAndFfmpeg, hasCurl, "curl"],
		[needsCurlAndFfmpeg, hasFfmpeg, "ffmpeg"]
	].forEach(([needsProgram, hasProgram, program]) => {
		if(needsProgram && !hasProgram) {
			console.error(`${scriptName}: ${program}: program missing`);
			process.exit(27);
		}
	});

	const downloads = playlist.download(cliInput.download.highEffort);

	let downloadsFinished = 0;

	function incrementFinishedDownload() {
		++downloadsFinished;

		if(downloadsFinished >= downloads.size) {
			console.log("All downloads finished");
		}
	}

	const onlyShowError: boolean = cliInput.download.onlyShowError;

	downloads.forEach((promise, basename) => {
		promise.then(
			() => {
				if(!onlyShowError) {
					console.log(`${basename}: success`);
				}
				incrementFinishedDownload();
			},
			() => {
				console.log(`${basename}: failure`);
				process.exitCode = 32;
				incrementFinishedDownload();
			}
		);
	});
}
